README
NAME
    Apache2::POST200 - Converting code 200 responses to POST requests to 302

SYNOPSIS
      LoadModule perl_module ...
      PerlLoadModule Apache2::POST200;

      POST200Storage dbi:mysql:db=db:localhost user password
      POST200Table p200 session data
      PerlOutputFilterHandler Apache2::POST200::Filter
      <Location "/-redirect-">
        SetHandler modperl
        PerlResponseHandler Apache2::POST200::Response
      </Location>

      RewriteEngine On

      # redirect GET/HEAD requests with a matching QUERY_STRING to
      # Apache2::POST200::Response
      RewriteCond %{REQUEST_METHOD} !=POST
      RewriteCond %{QUERY_STRING} ^-redirect-[A-Za-z0-9@=-]{32}$
      RewriteRule . /-redirect- [L,PT]

      # This is needed because some applications forget the
      # action="URL" attribute in their <form>s.
      # If so we get a POST request with a QUERY_STRING appended.
      # If it matches our pattern is must be cut off (the '?' in
      # the RewriteRule).
      RewriteCond %{REQUEST_METHOD} =POST
      RewriteCond %{QUERY_STRING} ^-redirect-[A-Za-z0-9@=-]{32}$
      RewriteRule (.+) $1? [N]

      # yes, it even works on a reverse proxy.
      RewriteRule ^/proxy/(.+) http://other.host.tld/$1 [P]

      # keep mod_alias (ScriptAlias) happy
      RewriteRule . - [PT]

     or

      # This is now a forward proxy setup.

      Listen localhost:8080
      <VirtualHost localhost:8080>
        POST200Storage dbi:mysql:db=db:localhost user password
        POST200Table p200 session data
        PerlOutputFilterHandler Apache2::POST200::Filter
        <Location "/-localhost-8080-">
          SetHandler modperl
          PerlResponseHandler Apache2::POST200::Response
        </Location>

        # defined a prefix instead of "-redirect-"
        POST200Location -localhost-8080-

        RewriteEngine On

        # redirect GET/HEAD requests with a matching QUERY_STRING to
        # Apache2::POST200::Response
        RewriteCond %{REQUEST_METHOD} !=POST
        RewriteCond %{QUERY_STRING} ^-localhost-8080-[A-Za-z0-9@=-]{32}$
        RewriteRule . /-localhost-8080- [L,PT]

        # This is needed because some applications forget the
        # action="URL" attribute in their <form>s.
        # If so we get a POST request with a QUERY_STRING appended.
        # If it matches our pattern is must be cut off (the '?' in
        # the RewriteRule). Then proxy it.
        RewriteCond %{REQUEST_METHOD} =POST
        RewriteCond %{THE_REQUEST} ^\w+\s(https?://[^\s\?]+)\?-localhost-8080-[A-Za-z0-9@=-]{32}(?:\s|$)
        RewriteRule . %1? [P]

        # Proxy all other. This is an alternative to the ProxyRequests
        # statement.
        RewriteCond %{THE_REQUEST} ^\w+\s(https?://\S+)
        RewriteRule . %1 [P]
      </VirtualHost>

INSTALLATION
     perl Makefile.PL
     make
     make test
     make install

DEPENDENCIES
    mod_perl2

DESCRIPTION
    A typical WEB application workflow is often similar to this:

      browser shows a form (1)
            |
            v
      user clicks submit (2)
            |
            v
      browser sends a POST request (3)
            |
            v
      server processes the form and replies with a temporary redirect (4)
            |
            v
      browser follows the redirect (5)
            |
            v
      server replies with HTTP code 200 (6)

    Steps 4 and 5 are necessary to let the user reload the page shown
    without having the server to reprocess the form.

    With this module the workflow is shortened from the point of view of the
    WEB server to this:

      browser shows a form (1)
            |
            v
      user clicks submit (2)
            |
            v
      browser sends a POST request (3)
            |
            v
      server processes the form and replies with HTTP code 200 (4)

    Apache2::POST200 intercepts the server reply, stores the response in a
    database and sends a temporary redirect to the browser. It also
    intercepts the following request from the browser and sends the stored
    reply.

  How it works
    This module inserts an request output filter that looks for replies for
    POST requests with a HTTP code of 200. If it finds one it saves the
    reply in a database and replaces the complete output with a temporary
    redirect (HTTP code 302) to the same URL but with a special marked query
    string appended.

    When the browser follows the redirect the module recognizes the query
    string and routes the request to its own response handler. The handler
    then reads the saved page from the database and sends it to the browser.

    Well, the request routing is actually done by a tricky translation
    handler such as mod_rewrite or Apache2::Translation.

    Note: the redirect must go to the same URL because some WEB application
    forget the "action" attribute in their "<form>" definitions.

  Configuration
    The module itself is loaded from the Apache configuration file via a
    "PerlLoadModule" directive. It then provides a few configuration
    directives of its own. All directives are allowed in server config,
    virtual host and directory contexts.

    Post200Storage dsn user password
        "Post200Storage" describes the database to be used. All 3 parameter
        are passed to the DBI::connect method, see DBI. User and password
        can be omitted if the database supports it.

        "Post200Storage None" disables the output filter. That means replies
        with a HTTP code 200 to a POST request are delivered as is.

    Post200Table table key-column data-column
        "Post200Table" describes the table to be used. The "key" column must
        be able to hold a 41-byte string of printable ascii characters. The
        key length may be extented in future versions of this module but a
        key will always consist of printable characters.

        For best performance create an index on the "key" column.

        The "data" column must be able to hold a variable size data block.
        The maximum size can be limited using "Post200DataBlockSize". If
        "Post200DataBlockSize" is not used the size completely depends on
        your response handlers. If possible use a BLOB type as "data"
        column.

        Although not used by the module it makes sense to add a 3rd column
        to the table. It should be a timestamp column with the default
        attribute set to "now()". Without it it's difficult to decide which
        records can be deleted.

        With a MySQL database a suitable table is created by:

         create table p200 (
           session varchar(50) primary key unique not null,
           data blob,
           tm timestamp not null default 'now'
         );
         create index p200_tm_idx on p200(tm);

        Deletion of expired pages is best done by a simple cron job, e.g.

         45 * * * * echo 'delete from p200 where now()-tm>3600' | mysql post200

    Post200Label marker
        By means of this marker the response handler recognizes a redirected
        request that it is responsible for. When the output filter generates
        a query string it starts with the "marker" as prefix.

        If omitted "-redirect-" is used.

        If the module is used on a forward proxy to repair external WEB
        applications choose a string here that is very likely to be used
        only by your proxy.

    Post200Secret secret initvector
        To make sure the key provided by the browser via the query string
        was generated by the filter it is encrypted. "secret" and
        "initvector" are arbitrary strings, see Crypt::Blowfish.

        If omitted 2 strings are used that once came out of /dev/random on
        my box.

    Post200IpCheck On|Off
        With this directive set the response handler sends a page only to
        the same client where the redirect was sent to. This prevents that
        redirected URLs are mailed around as long as the 2 clients are not
        connected through the same proxy.

        Default is On.

    Post200DataBlockSize Bytes
        This directive defines the maximum size of a data item written to
        the database.

        If omitted the blocksize depends on the response handler.

Some Considerations
  Simple sessions
    One way to look at this module is that it provides some simple session.
    Often a WEB application is simply a collection of forms gathering some
    information. And only after the last form is filled out all of it is to
    be written to a data store.

    With "Apache2::POST200" you can save the information gathered so far in
    hidden fields rather than saving them in a session structure at server
    side.

    Well, the database that this module uses is such a session structure at
    server side, but forget about the internals for now. Look at it from
    another level, ;-).

  Simpler developement
    "Apache2::POST200" frees your application from handling the refreshing
    problem by its own. Hence, developement becomes simpler.

  Performance
    Using "Apache2::POST200" might even lead to a performance gain. Often
    the application logic is almost the same (authentication, authorization,
    fetching application data and so on) be it for the generation of the
    Location header or for the redirected request. With this module these
    checks are done only once.

  Frontend / Backend
    Here comes the most valuable point. WEB applications often use a
    frontend / backend setup, where the backend serves the dynamic content.
    The frontend is a lightweight WEB server that serves static content and
    delegates request for dynamic content as a proxy to the backend.

    This setup is chosen because generating dynamic content often leads to
    very memory consuming processes and, thus, the necessity to limit their
    number. But such a setup does not forbid using mod_perl also in the
    frontend. Only the memory consumption must be small and limited. Since
    "Apache2::POST200" reads the data from the database one block at a time
    it meets this condition.

    Now look at what happens without "Apache2::POST200". The initial POST
    request as well as the redirected GET occupy a frontend and a backend
    instance. If "Apache2::POST200" is used at the frontend only the initial
    POST involves both frontend and backend. The subsequent GET is handled
    exclusively by the frontend.

    "Apache2::POST200" may be used even at frontend and backend. The output
    filter runs at the backend inserting data in the database. The response
    handler runs at the frontend. Limit the block size in this case with
    "Post200DataBlockSize" to avoid bloating the frontend.

  Repair external WEB applications
    Another useful application of "Apache2::POST200" is to repair external
    applications. Suppose you have a WEB application written in some closed
    language or running on an external server that emits code 200 replies to
    POST requests. Setup a reverse or even forward proxy with
    "Apache2::POST200". It repairs the application without having access to
    it.

TODO
    Caching
        With a keep-alive connection the redirected request is most likely
        to come in over the same connection. Thus, some caching in a
        connection pnote would be good.

    User check
        Client IP checking may not be sufficient. The filter could check
        "$r->user", set a special flag in the redirect param and save the
        user name. The rewrite rules could then check the flag and require a
        valid-user. Then the response handler can verify "$r->user" against
        the saved user.

SEE ALSO
    Apache2::Translation

AUTHOR
    Torsten Foertsch, <torsten.foertsch@gmx.net>

SPONSORING
    Sincere thanks to Arvato Direct Services (http://www.arvato.com/) for
    sponsoring this module.

COPYRIGHT AND LICENSE
    Copyright (C) 2005 by Torsten Foertsch

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Apache2::POST302.

0.01  Tue Dec  6 19:12:08 2005
	- original version; created by h2xs 1.23 with options
		-A -b 5.8.0 -n Apache2::POST302 -XS



t/TESTlib/mp.pm
package TESTlib::mp;

use strict;
use Apache2::Const -compile=>qw{OK DECLINED};
use Apache2::ServerRec;
use Apache2::ServerUtil;
use Apache2::RequestRec;
use Apache2::RequestIO;
use Apache2::Connection;

sub method {
  my $r=shift;

  $r->content_type('text/slain');
  $r->headers_out->{'X-My-Header'}='hallo opi';
  $r->err_headers_out->{'X-My-Error'}='error';
  $r->print($r->method.':'.$r->args);

  return Apache2::Const::OK;
}

sub chunks {
  my $r=shift;

  $r->content_type('text/slain');
  $r->print('x' x 1);
  $r->rflush;
  $r->print('x' x 2);
  $r->rflush;
  $r->print('x' x 3);

  return Apache2::Const::OK;
}

sub big {
  my $r=shift;

  $r->content_type('text/slain');
  for( 1..1024 ) {
    $r->print('y' x 10240);
    $r->rflush;
  }

  return Apache2::Const::OK;
}

sub proxy {
  my $r=shift;

  my $backend=Apache2::ServerUtil->server->next;
  $backend=$backend->server_hostname.':'.$backend->port;
  my $url=$r->uri;
  $url=~s!^/+[^/]+!!;
  $url="http://$backend/$url";
  $r->proxyreq(2);
  $r->filename("proxy:$url");
  $r->handler('proxy-server');

  return Apache2::Const::DECLINED;
}

sub setip {
  my $r=shift;

  $r->connection->remote_ip('85.25.96.79');

  return Apache2::Const::DECLINED;
}

1;


lib/Apache2/POST200.pm
package Apache2::POST200;

use 5.008;
use strict;
use warnings;
no warnings qw(uninitialized);

use Apache2::RequestRec;
use Apache2::RequestUtil;
use Apache2::RequestIO;
use Apache2::ServerUtil;
use Apache2::Connection;
use Apache2::CmdParms;
use Apache2::Module;
use Apache2::Filter;
use APR::Brigade;
use APR::Bucket;
use APR::Table;
use Apache2::Const -compile=>qw{OK DECLINED
				TAKE1 TAKE12 TAKE123 TAKE3 FLAG OR_ALL
				M_POST M_GET
				HTTP_OK REDIRECT NOT_FOUND};

use MIME::Base64 ();
use Crypt::CBC ();
use Crypt::Blowfish ();
use Digest::MD5 ();
use Digest::CRC ();
use DBI;

our $VERSION = '0.03';
my $rcounter=0;

# these 2 values were once read from /dev/random on my box
my $default_key=("tFS\343x\314\357uh\212W\177+#\332\0q\317S\231\321\316\270H".
		 "\252\205\313\264\357LT\16h\362\36\354cK\317\362\e\253`[8".
		 "\211\365\347\217:\f1\224\321L*");
my $default_iv="P\363\32\310\24\340\265\373";

my $msg302=<<'EOF';
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>302 Found</title>
</head><body>
<h1>Found</h1>
<p>The document has moved <a href="%{location}">here</a>.</p>
</body></html>
EOF

my @directives=
  (
   {
    name         => 'Post200Storage',
    func         => __PACKAGE__ . '::config123',
    req_override => Apache2::Const::OR_ALL,
    args_how     => Apache2::Const::TAKE123,
    errmsg       => 'Post200Storage DBI-DSN [USER] [PASSWORD]',
    cmd_data     => 'storage',
   },
   {
    name         => 'Post200Table',
    func         => __PACKAGE__ . '::config123',
    req_override => Apache2::Const::OR_ALL,
    args_how     => Apache2::Const::TAKE3,
    errmsg       => 'Post200Table TABLENAME KEY-COLUMN VALUE-COLUMN',
    cmd_data     => 'table',
   },
   {
    name         => 'Post200Label',
    func         => __PACKAGE__ . '::config123',
    req_override => Apache2::Const::OR_ALL,
    args_how     => Apache2::Const::TAKE1,
    errmsg       => 'Post200Label marker (default: "-redirect-")',
    cmd_data     => 'location',
   },
   {
    name         => 'Post200Secret',
    func         => __PACKAGE__ . '::config123',
    req_override => Apache2::Const::OR_ALL,
    args_how     => Apache2::Const::TAKE12,
    errmsg       => 'Post200Secret SECRET [INITVECTOR]',
    cmd_data     => 'secret',
   },
   {
    name         => 'Post200IpCheck',
    func         => __PACKAGE__ . '::config123',
    req_override => Apache2::Const::OR_ALL,
    args_how     => Apache2::Const::FLAG,
    errmsg       => 'Post200IpCheck On|Off (default: On)',
    cmd_data     => 'checkip',
   },
   {
    name         => 'Post200DataBlockSize',
    func         => __PACKAGE__ . '::config123',
    req_override => Apache2::Const::OR_ALL,
    args_how     => Apache2::Const::TAKE1,
    errmsg       => 'Post200DataBlockSize bytes',
    cmd_data     => 'blocksize',
   },
  );
Apache2::Module::add(__PACKAGE__, \@directives);

my %extra_config=
  (
   secret=>sub {
     unless( length $_[0]->[1] ) {
       $_[0]->[1]='hex:'.unpack( 'H*', $default_iv );
     }
     map {
       if( /^hex:(.+)/ ) {
	 $_=pack( 'H*', $_ );
	 $_=substr( $_, 0, 56 ) if( length($_)>56 );
       } elsif( /^b64:(.+)/ ) {
	 $_=MIME::Base64::decode_base64( $_ );
	 $_=substr( $_, 0, 56 ) if( length($_)>56 );
       } else {
	 $_=Digest::MD5::md5( $_ );
       }
     } @{$_[0]};
   },
  );

sub config123 {
  my($I, $parms, @args)=@_;
  $I->{$parms->info}=[@args[0..2]];
  $extra_config{$parms->info}->( $I->{$parms->info} )
    if( exists $extra_config{$parms->info} );
}

sub DIR_CREATE {
  my ($class, $parms)=@_;
  return bless {
		secret=>[$default_key, $default_iv],
		location=>['-redirect-'],
		checkip=>['1'],
	       } => $class;
}

sub DIR_MERGE {
  my ($base, $add) = @_;

  my %new=(%$base, %$add);

  return bless \%new, ref($base);
}

sub Response {
  my $r=shift;

  my $cf=Apache2::Module::get_config(__PACKAGE__, $r->server,
				     $r->per_dir_config);

  return Apache2::Const::NOT_FOUND
    unless( $r->method_number==Apache2::Const::M_GET and
	    length( $r->args )==32+length($cf->{location}->[0]) );

  my $crypt=Crypt::CBC->new( {key=>$cf->{secret}->[0],
			      cipher=>'Blowfish',
			      iv=>$cf->{secret}->[1],
			      regenerate_key=>0,
			      prepend_iv=>0} );

  my $session=$r->args;
  $session=~s/^\Q$cf->{location}->[0]\E//;
  my $db_key=$session;
  $session=~tr[@\-][+/];
  $session=$crypt->decrypt( MIME::Base64::decode_base64( $session ) );

  my $crc=Digest::CRC::crc8( substr( $session, 1 ) );

  my ($crc2, undef, undef, undef, undef, @ip)=unpack 'CNNnNC8', $session;

  unless( $crc==$crc2 ) {
    $r->warn( __PACKAGE__.": CRC checksum error" );
    return Apache2::Const::NOT_FOUND;
  }

  if( $cf->{checkip}->[0] and join('.', @ip[0..3]) ne $r->connection->remote_ip ) {
    $r->warn( __PACKAGE__.": IP check failed" );
    return Apache2::Const::NOT_FOUND;
  }

  my $dbh=DBI->connect( @{$cf->{storage}}[0..2],
			{
			 AutoCommit=>1,
			 PrintError=>0,
			 RaiseError=>0,
			} )
    or do {
      $r->warn( "Cannot connect to $cf->{storage}->[0]: $DBI::errstr" );
      return Apache2::Const::NOT_FOUND;
    };

  my $stmt=$dbh->prepare("SELECT $cf->{table}->[1], $cf->{table}->[2] ".
			 "FROM $cf->{table}->[0] ".
			 "WHERE $cf->{table}->[1] LIKE ? ".
			 "ORDER BY $cf->{table}->[1] ASC")
    or do {
      $r->warn( "Cannot prepare SELECT statement: ".$dbh->errstr );
      $dbh->disconnect;
      return Apache2::Const::NOT_FOUND;
    };

  $session=$db_key;
  $stmt->execute( $session.':%' )
    or do {
      $r->warn( "Cannot execute SELECT statement: ".$dbh->errstr );
      $dbh->disconnect;
      return Apache2::Const::NOT_FOUND;
    };

  my $i=1;
  while( my $l=$stmt->fetchrow_arrayref ) {
    if( $l->[0] eq sprintf( '%s:%08d', $session, $i ) ) {
      if( $i==1 ) {		# headers_out
	$r->headers_out->clear;
	foreach my $line (split /\n/, $l->[1]) {
	  $r->headers_out->add(split /: /, $line, 2)
	    if( length $line );
	}
      } elsif( $i==2 ) {	# err_headers_out
	$r->err_headers_out->clear;
	foreach my $line (split /\n/, $l->[1]) {
	  $r->err_headers_out->add(split /: /, $line, 2)
	    if( length $line );
	}
      } elsif( $i==3 ) {	# content-type
	$r->content_type($l->[1]);
      } else {			# data
	$r->print( $l->[1] );
      }
    } else {
      $r->warn( "Read incomplete data from database" );
    }
    $i++;
  }

  return Apache2::Const::OK;
}

sub Filter {
  my ($f, $bb) = @_;

  unless( $f->ctx ) {
    my $r=$f->r;

    my $cf=Apache2::Module::get_config(__PACKAGE__, $r->server,
				       $r->per_dir_config);

    if( $r->main or		# skip filtering for subrequests
	$r->method_number!=Apache2::Const::M_POST or
	!(do{no warnings 'numeric';$r->status_line==Apache2::Const::HTTP_OK} or
	  !length( $r->status_line ) && $r->status==Apache2::Const::HTTP_OK) or
	!exists($cf->{storage}) or
	lc $cf->{storage}->[0] eq 'none' or
	!exists($cf->{table}) or
	lc $cf->{table}->[0] eq 'none') {
      $f->remove;
      return Apache2::Const::DECLINED;
    }

    my $session=pack( 'NNnNC8',
		      $r->request_time, $$, $rcounter++,
		      $r->connection->id,
		      split( /\./, $r->connection->remote_ip, 4 ),
		      split( /\./, $r->connection->local_ip, 4 ),
		    );
    $rcounter%=2**16;

    $session=pack( 'C', Digest::CRC::crc8( $session ) ).$session;

    my $crypt=Crypt::CBC->new( {key=>$cf->{secret}->[0],
				cipher=>'Blowfish',
				iv=>$cf->{secret}->[1],
				regenerate_key=>0,
				prepend_iv=>0} );

    $session=MIME::Base64::encode_base64( $crypt->encrypt( $session ), '' );

    # The Base64 Alphabet consists of [A-Za-z0-9+/] where each character
    # represents 6 bits (0-64) plus the equal sign (=) as padding character
    # To get a valid URI part [+/] must be avoided since they have special
    # meaning in URIs. We change them to [@-].
    # Thus, the resulting alphabet contains neither [/#?+] nor [_%]. The
    # former are dangerous in URIs the latter in SQL LIKE statements.
    $session=~tr[+/][@\-];

    my $dbh=DBI->connect( @{$cf->{storage}}[0..2],
			  {
			   AutoCommit=>1,
			   PrintError=>0,
			   RaiseError=>0,
			  } )
      or do {
	$r->warn( "Cannot connect to $cf->{storage}->[0]: $DBI::errstr" );
	$f->remove;
	return Apache2::Const::DECLINED;
      };

    $dbh->begin_work;

    my $headers='';
    $r->headers_out->do(sub{$headers.="$_[0]: $_[1]\n";1;});
    my $err_headers='';
    $r->err_headers_out->do(sub{$err_headers.="$_[0]: $_[1]\n";1;});

    # check if the table exists and can be written
    my $stmt=$dbh->prepare("INSERT INTO $cf->{table}->[0] ".
			   "($cf->{table}->[1], $cf->{table}->[2]) ".
			   "VALUES (?, ?)")
      or do {
	$r->warn( "Cannot prepare INSERT statement: ".$dbh->errstr );
	$dbh->disconnect;
	$f->remove;
	return Apache2::Const::DECLINED;
      };

    $stmt->execute( $session.':00000001', $headers) &&
    $stmt->execute( $session.':00000002', $err_headers) &&
    $stmt->execute( $session.':00000003', $r->content_type)
      or do {
	$r->warn( "Cannot insert into $cf->{table}->[0]: ".$dbh->errstr );
	$dbh->disconnect;
	$f->remove;
	return Apache2::Const::DECLINED;
      };

    my $loc=$r->the_request;	# don't count on $r->uri or $r->unparsed_uri
				# they may have been changed
    $loc=~s/^\w+\s//;		# strip "POST " at head
    $loc=~s/[?#\s].*//;		# strip any parameters, anchor and "HTTP/1.1"
    unless( $loc=~m!^https?://! ) { # can already be for proxy requests
      my $proto=(($r->can('is_https') && $r->is_https or # using Apache::SSLLookup
		  $r->connection->can('is_https') && $r->connection->is_https or # Apache2::ModSSL
		  $r->subprocess_env('HTTPS'))
		 ? 'https'
		 : 'http');
      my $port=':'.$r->get_server_port;
      $port='' if( $port eq ':80' && $proto eq 'http' or
		   $port eq ':443' && $proto eq 'https' );
      $loc=$proto.'://'.$r->get_server_name.$port.$loc;
    }
    $loc.='?'.$cf->{location}->[0].$session;

    my $msg=$msg302;
    $msg=~s/%{location}/$loc/g;

    $r->status( Apache2::Const::REDIRECT );
    $r->status_line( Apache2::RequestUtil::get_status_line(Apache2::Const::REDIRECT) );
    $r->headers_out->clear;
    $r->err_headers_out->clear;
    $r->content_type( 'text/html; charset=iso-8859-1' );
    $r->headers_out->{'Content-Length'}=length $msg;
    $r->headers_out->{'Location'}=$loc;

    $f->ctx( {
	      dbh=>$dbh,
	      stmt=>$stmt,
	      session=>$session,
	      nr=>4,
	      msg=>$msg,
	      bs=>$cf->{blocksize}->[0],
	     } );
  }

  my $ctx=$f->ctx;
  while (my $e = $bb->first) {
    if( $e->is_eos ) {
      $ctx->{dbh}->commit;
      $e->remove;
      my $bbnew=APR::Brigade->new( $f->c->pool, $f->c->bucket_alloc );
      $bbnew->insert_tail(APR::Bucket->new( $bbnew->bucket_alloc, $ctx->{msg} ));
      $bbnew->insert_tail($e);
      $f->next->pass_brigade( $bbnew );
    } else {
      $e->read(my $buf);
      if( length $buf ) {
	if( $ctx->{bs}>0 ) {
	  my ($i, $len, $bs)=(0, length( $buf ), $ctx->{bs});
	  while( $i<$len ) {
	    $ctx->{stmt}->execute
	      ( sprintf( '%s:%08d', $ctx->{session}, $ctx->{nr}++ ),
		substr( $buf, $i, $bs ) );
	    $i+=$bs;
	  }
	} else {
	  $ctx->{stmt}->execute
	    ( sprintf( '%s:%08d', $ctx->{session}, $ctx->{nr}++ ), $buf );
	}
      }
      $e->delete;
    }
  }

  return Apache2::Const::OK;
}

1;
